### PR TITLE
Feature create plants

### PR DESCRIPTION
**Pull request by SabinePfanner**
Quite a complex ticket, therefore some remarks:

1. We've created a component `Modal` and `ToastMessage` to handle the button actions e.g. dismiss, save etc.
2. We've created a reusable `button` component
3. We've create a `form` component with several custom inputs e.g. `select`, `radio button`, `range scale`
4. The form does on purpose not include the functionality to add a plant image (cf. user story), therefore a placeholder image is picked for new plants for now.
5. We've implemented a page `/create` to handle the creation of a new plant. If the user saves the new crop, there is a router push including a parameter to handle the toast message afterwards

US: https://github.com/users/SabinePfanner/projects/2/views/1?pane=issue&itemId=64301835

### CODE REVIEW

File: components/CardList.js [hunk 1]
```diff
@@ -17,21 +18,23 @@ const StyledListElement = styled.li`
...
+    <>
+      <StyledList>
+        {plants.map((plant) => {
+          return (
+            <StyledListElement key={plant._id}>
+              <Card
+                name={plant.name}
+                cropType={plant.cropType}
+                image={plant.image}
+                isFavorite={favoriteIDs.includes(plant._id) ? true : false}
```
Comment (doemser) [inline components/CardList.js]:
```suggestion
                isFavorite={favoriteIDs.includes(plant._id)}
```

File: components/Form.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+const RangeInputLabels = styled.div`
+  display: flex;
+  justify-content: space-between;
+`;
+
+export default function Form({ onSubmit, onDismiss }) {
+  // Select Crop Type
+
+  const cropTypes = ["Fruit", "Herb", "Vegetable", "Other"];
```
Comment (doemser) [inline components/Form.js]: There is no need to place it inside the component

File: components/Form.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+  const [currentCropType, setCurrentCropType] = useState(null);
+
+  function handleCropTypeChange(value) {
+    setCurrentCropType(value);
+  }
+
+  // Select Placement
+
+  const placements = ["Bed", "Pot", "Pot or Bed"];
```
Comment (doemser) [inline components/Form.js]: same here

File: components/Form.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+  const [currentPlacement, setCurrentPlacement] = useState(null);
+
+  function handlePlacementChange(value) {
+    setCurrentPlacement(value);
+  }
+
+  // Select Growing Conditions
+
+  const growingConditions = ["Sunny", "Partial Shade"];
```
Comment (doemser) [inline components/Form.js]: and here

File: components/Form.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    ) {
+      alert("Please fill out all inputs");
+      return;
+    } else {
+      formData.set("cropType", currentCropType);
+      formData.set("placement", currentPlacement);
+      formData.set("growingConditions", currentGrowingConditions);
+    }
+    const plantData = Object.fromEntries(formData);
+    onSubmit(plantData);
```
Comment (doemser) [inline components/Form.js]:
```suggestion
    if (
      currentCropType === null ||
      currentPlacement === null ||
      currentGrowingConditions === null
    ) {
      alert("Please fill out all inputs");
    } else {
      formData.set("cropType", currentCropType);
      formData.set("placement", currentPlacement);
      formData.set("growingConditions", currentGrowingConditions);
      const plantData = Object.fromEntries(formData);
      onSubmit(plantData);
    }
  
```

File: components/ModalAndToast/Modal.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      timeoutId = setTimeout(() => {
+        setActionConfirmed(false);
+        onModalOpen();
+        router.push(`/`);
+      }, 3000);
+    }
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  });
```
Comment (doemser) [inline components/ModalAndToast/Modal.js]: Every usEffect needs a dependency array, not having one is almost the same as not using an useEffect

File: components/ModalAndToast/ToastMessage.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  return (
+    <>
+      <Background>
+        <ModalBox $isActionConfirmed={true}>
+          <ModalContent>
+            <ModalInfo>{toastMessageText}</ModalInfo>
+          </ModalContent>
+        </ModalBox>
+      </Background>
+    </>
```
Comment (doemser) [inline components/ModalAndToast/ToastMessage.js]: you can remove the fragments

File: components/StyledElements/CreateEditDelete.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    transform: scale(1.1);
+    title: "createCrop";
+  }
+`;
+
+export default function CreateNewPlantButton() {
+  return (
+    <>
+      <StyledLink href="/create">
+        <SvgIcon variant={"plus"} color={"#E23D28"} title={"createCrop"} />
```
Comment (doemser) [inline components/StyledElements/CreateEditDelete.js]:
```suggestion
        <SvgIcon variant="plus" color="#E23D28" title="createCrop" />
```

File: components/StyledElements/Select.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    &:hover, :focus, :focus:hover {
+      background-color: var(--color-green);
+      color: #fff;
+      outline: 0;
+    }
+  `;
+  return (
+    <>
+      <SelectLabelButton onClick={handleToggle}>
+        {labelButtonText !== "" ? labelButtonText : label}
```
Comment (doemser) [inline components/StyledElements/Select.js]:
```suggestion
        {labelButtonText ? labelButtonText : label}
```

File: pages/[id].js [hunk 10]
```diff
@@ -15,10 +18,33 @@ const StyledLink = styled(Link)`
 `;
 
 export default function DetailsPage({ favoriteIDs, onToggleFavorite }) {
+  const router = useRouter();
+  const query = router.query;
+  console.log(query);
```
Comment (doemser) [inline pages/[id].js]: remove unused console.logs

File: pages/[id].js [hunk 11]
```diff
@@ -15,10 +18,33 @@ const StyledLink = styled(Link)`
...
+  const [showToastMessage, setShowToastMessage] = useState(true);
+
+  useEffect(() => {
+    let timeoutId;
+    if (query.isnew !== undefined) {
```
Comment (doemser) [inline pages/[id].js]:
```suggestion
    if (query.isnew) {
```

File: pages/[id].js [hunk 12]
```diff
@@ -15,10 +18,33 @@ const StyledLink = styled(Link)`
...
+
   return (
     <>
       <StyledLink href="/">←</StyledLink>
-      <PlantDetails favoriteIDs={favoriteIDs} onToggleFavorite={onToggleFavorite}/>
+      <PlantDetails
+        favoriteIDs={favoriteIDs}
+        onToggleFavorite={onToggleFavorite}
+      />
+      {query.isnew !== undefined && showToastMessage && (
```
Comment (doemser) [inline pages/[id].js]:
```suggestion
      {query.isnew && showToastMessage && (
```

File: pages/api/plants/index.js [hunk 13]
```diff
@@ -7,6 +7,19 @@ export default async function handler(request, response) {
...
     return response.status(200).json(plants);
+  }
+  if (request.method === "POST") {
+    try {
+      const plantData = request.body;
+      const newPlant = await Plant.create(plantData);
+      response
+        .status(201)
+        .json({ message: "New crop successfully created", id: newPlant._id });
+      console.log(newPlant);
```
Comment (doemser) [inline pages/api/plants/index.js]: remove unused console.logs

File: pages/api/plants/index.js [hunk 14]
```diff
@@ -7,6 +7,19 @@ export default async function handler(request, response) {
...
+    } catch (error) {
+      console.log(error);
```
Comment (doemser) [inline pages/api/plants/index.js]: this console log I would consider valid though

File: pages/create.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+      console.error(response.status);
+      return;
+    }
+    if (response.ok) {
+      const responseObject = await response.json();
+      console.log(response.status);
+      mutate();
+      router.push(`/${responseObject.id}?isnew`);
+    }
+  }
```
Comment (doemser) [inline pages/create.js]:
```suggestion
    if (!response.ok) {
      console.error(response.status);
    } else {
      const responseObject = await response.json();
      mutate();
      router.push(`/${responseObject.id}?isnew`);
    }
  }
```

**Review by doemser — CHANGES_REQUESTED** [review]
Good Job! Just some minor changes I request.

Make sure people are extensively trying your custom dropdowns in QA

File: pages/create.js (hunk 15 above)
Comment (doemser) [reply pages/create.js]: also, do you really need to mutate if you push to another route?

File: pages/[id].js (hunk 11 above)
Comment (SabinePfanner) [reply pages/[id].js]: We've implemented it this way, as the router push was only to `?isnew` without any value behind the key. To implement your suggested change, we therefore now push to route `?isnew=true` to make it work.

File: pages/api/plants/index.js (hunk 14 above)
Comment (SabinePfanner) [reply pages/api/plants/index.js]: We've removed it as well, as not needed for us and `response.status` is already logged.

File: components/ModalAndToast/Modal.js [hunk 16]
```diff
@@ -37,34 +37,40 @@ export default function Modal({
...
-                type="button"
-                onClick={handleConfirm}
-              >
-                {confirmButtonLabel}
-              </StyledButton>
-            </ButtonGroup>
-          </ModalContent>
-        </ModalBox>
-      </Background>
+      {!actionConfirmed && (
```
Comment (doemser) [inline components/ModalAndToast/Modal.js]: you could go for a simple if/else

```javascript
if(actionConfirmed) {
 return <Toast/>
} else {
 return <Modal/>
}
```

File: components/Form.js [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+  function handleSubmit(event) {
+    event.preventDefault();
+    const formData = new FormData(event.target);
+    // Handle custom selects: if not null, add data to form data, otherwise show a custom alert message
+    if (
+      currentCropType === null ||
+      currentPlacement === null ||
+      currentGrowingConditions === null
+    ) {
+      alert("Please fill out all inputs");
```
Comment (Jonstar22) [inline components/Form.js:97]: You dont have to write that, you can just use `required` but if you want it to be a `alert` its also fine :)

File: components/PlantDetails.js [hunk 18]
```diff
@@ -69,9 +69,21 @@ export default function PlantDetails({ favoriteIDs, onToggleFavorite }) {
...
         </StyledListElement>
-        <StyledListElement>Water demand: {plant.waterDemand}</StyledListElement>
         <StyledListElement>
-          Nutrient demand: {plant.nutrientDemand}
+          Water demand:{" "}
+          {plant.waterDemand === "1"
+            ? "Low"
+            : plant.nutrientDemand === "2"
+            ? "Medium"
+            : "High"}
```
Comment (Jonas-Wendland) [inline components/PlantDetails.js:78]: im not sure but isnt there an easier way ? maybe a switch case function?

File: components/StyledElements/CreateEditDelete.js (hunk 8 above)
Comment (Jonas-Wendland) [inline components/StyledElements/CreateEditDelete.js]: What Sir doemser said!

File: components/StyledElements/RangeSlider.js [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+const makeLongShadow = (color, size) => {
+  let i = 1;
+  let shadow = `${i}px 0 0 ${size} ${color}`;
+
+  for (; i < 700; i++) {
+    shadow = `${shadow}, ${i}px 0 0 ${size} ${color}`;
+  }
+
+  return shadow;
+};
```
Comment (Jonas-Wendland) [inline components/StyledElements/RangeSlider.js:24]: I do not 100% understand what this is for ?  Is it for the slider to change color when slided ?

**Review by Jonas-Wendland — COMMENTED** [review]
Once again, there's a lot of styling involved :3. However, one thing that makes reviewing the code challenging is that the styling components are placed between the logic and the return statement. Personally, I prefer having them below the code to keep things more distinct.

Also a huge ticket maybe minimize the next one in smaller steps

Overall, it's quite good, but please pay attention to the comments from Sergeant Doemser!

File: components/PlantDetails.js (hunk 18 above)
Comment (SabinePfanner) [reply components/PlantDetails.js:78]: Talked to Philip about this suggestion and he mentioned that a switch is not possible in return / jsx. Therefore we first need to extract the function and call it in the return afterwards. As this is currently just a work around and might be changed while focusing on styling and optic, we would therefore prefer to keep it like this for the moment.

File: components/StyledElements/RangeSlider.js (hunk 19 above)
Comment (SabinePfanner) [reply components/StyledElements/RangeSlider.js:24]: Discussed directly.

**Pull request merged**